framework/aster-frame/src/task/scheduler.rs

```diff
@@ -1,64 +1,146 @@
 // SPDX-License-Identifier: MPL-2.0
 
+use crate::config::SCHED_DEBUG_LOG;
 use crate::prelude::*;
-use crate::sync::SpinLock;
-use crate::task::Task;
+use crate::sync::{SpinLock, SpinLockGuard};
+use crate::task::{NeedResched, ReadPriority, Task};
 
-use lazy_static::lazy_static;
+use log::debug;
 
-lazy_static! {
-    pub(crate) static ref GLOBAL_SCHEDULER: SpinLock<GlobalScheduler> =
-        SpinLock::new(GlobalScheduler { scheduler: None });
+pub(crate) static GLOBAL_SCHEDULER: spin::Once<SpinLock<GlobalScheduler>> = spin::Once::new();
+
+pub fn init() {
+    GLOBAL_SCHEDULER.call_once(|| SpinLock::new(GlobalScheduler { scheduler: None }));
 }
 
 /// A scheduler for tasks.
 ///
-/// An implementation of scheduler can attach scheduler-related information
-/// with the `TypeMap` returned from `task.data()`.
-pub trait Scheduler: Sync + Send {
-    fn enqueue(&self, task: Arc<Task>);
+/// Operations on the scheduler should be performed with interrupts disabled,
+/// which has been ensured by the callers of the `GLOBAL_SCHEDULER`.
+/// Therefore, implementations of this trait do not need to worry about interrupt safety.
+pub trait Scheduler<T: NeedResched + ReadPriority = Task>: Sync + Send {
+    /// Called when a task enters a runnable state.
+    /// Add the task to the scheduler.
+    fn enqueue(&self, task: Arc<T>);
+
+    /// Called when the task is no longer alive.
+    /// Remove the task-related from the scheduler.
+    ///
+    /// A similar method in the linux scheduler is `dequeue_task`.
+    ///
+    /// Return `true` if the task was in the scheduler.
+    fn remove(&self, task: &Arc<T>) -> bool;
+
+    /// Choose the most appropriate task eligible to run next.
+    fn pick_next_task(&self) -> Option<Arc<T>>;
 
-    fn dequeue(&self) -> Option<Arc<Task>>;
+    /// Tells whether the given current task should be preempted
+    /// by other tasks in the queue.
+    fn should_preempt(&self, cur_task: &Arc<T>) -> bool;
 
-    /// Tells whether the given task should be preempted by other tasks in the queue.
-    fn should_preempt(&self, task: &Arc<Task>) -> bool;
+    /// Handle a tick from the timer.
+    /// Modify the states of the given task(the current task in `PROCESSOR`)
+    /// according to the time update.
+    ///
+    /// # Arguments
+    ///
+    /// * `cur_task` - the task to be charged, must be held by the processor, and not in runqueue
+    fn tick(&self, cur_task: &Arc<T>);
+
+    /// Modify states before yielding the current task.
+    /// Set the `need_resched` flag of the current task.
+    fn before_yield(&self, cur_task: &Arc<T>) {
```

Comment:
Based on the current code, it appears that the `before_yield` method solely modifies the task state without any direct association with the `Scheduler` trait. So why `before_yield` should be a method of the `Scheduler` trait?

Furthermore, it is recommended to prefix all methods with a verb. For instance, you could consider renaming it to `prepare_task_for_yield` to adhere to this naming convention.

Reply:
The method is required in `Scheduler` trait by the other scheduler impls. O(1) scheduler needs to update some fields of the current task held by the processor before `schedule` in `yield_now`.
The method name will be changed.

Reply:
Ok, we can currently keep the method.